src/index.test.ts

```diff
@@ -179,4 +184,42 @@ describe('retry logic in middleware connected to a port', () => {
 
     expect(messages).toHaveLength(5);
   });
+
+  it('throw error when requests are retried more than 3 times', async () => {
```

Comment:
Nit: Maybe we should have a test for the case where a request has no id?

Reply:
👍 I added that